src/flow.jl

```diff
@@ -720,24 +722,385 @@ function update(
+        mannings_n = n_land,
+        volume = zeros(n),
+        error = zeros(n),
+        runoff = zeros(n),
+        h = zeros(n),
+        h_av = zeros(n),
+        z = elevation,
+        froude_limit = froude_limit,
+        rivercells = river,
+    )
+
+    return sw_land, indices
+end
+
 """
     stable_timestep(sw::ShallowWaterRiver)
+    stable_timestep(sw::ShallowWaterLand)
 
-Compute a stable timestep size for the local inertial approach for river flow, based on
-Bates et al. (2010).
+Compute a stable timestep size for the local inertial approach, based on Bates et al. (2010).
 
 Δt = α * (Δx / sqrt(g max(h))
 """
-function stable_timestep(sw::ShallowWaterRiver)
-    Δtₘᵢₙ = Inf
+function stable_timestep(sw::ShallowWaterRiver{T})::T where {T}
+    Δtₘᵢₙ = T(Inf)
     for i = 1:sw.n
         Δt = sw.α * sw.dl[i] / sqrt(sw.g * sw.h[i])
         Δtₘᵢₙ = Δt < Δtₘᵢₙ ? Δt : Δtₘᵢₙ
     end
-    Δtₘᵢₙ = isinf(Δtₘᵢₙ) ? 10.0 : Δtₘᵢₙ
+    Δtₘᵢₙ = isinf(Δtₘᵢₙ) ? T(10.0) : Δtₘᵢₙ
+    return Δtₘᵢₙ
+end
+
+function stable_timestep(sw::ShallowWaterLand{T})::T where {T}
+    Δtₘᵢₙ = T(Inf)
+    for i = 1:sw.n
+        if !sw.rivercells[i]
+            Δt = sw.α * min(sw.xl[i], sw.yl[i]) / sqrt(sw.g * sw.h[i])
+            Δtₘᵢₙ = Δt < Δtₘᵢₙ ? Δt : Δtₘᵢₙ
+        end
+    end
+    Δtₘᵢₙ = isinf(Δtₘᵢₙ) ? T(10.0) : Δtₘᵢₙ
     return Δtₘᵢₙ
 end
 
+function update(
+    sw::ShallowWaterLand{T},
+    swr::ShallowWaterRiver{T},
+    network;
+    inflow_wb = nothing,
+    doy = 0,
+    update_h = false,
+) where {T}
+
+    @unpack nodes_at_link, links_at_node = network.river
+
+    inds_riv = network.index_river
+
+    if !isnothing(swr.reservoir)
+        swr.reservoir.inflow .= 0.0
+        swr.reservoir.totaloutflow .= 0.0
+    end
+    if !isnothing(swr.lake)
+        swr.lake.inflow .= 0.0
+        swr.lake.totaloutflow .= 0.0
+    end
+    swr.q_av .= 0.0
+    swr.h_av .= 0.0
+
+    t = T(0.0)
+    while t < swr.Δt
+        Δt_river = stable_timestep(swr)
+        Δt_land = stable_timestep(sw)
+        Δt = min(Δt_river, Δt_land)
+        shallowwater_river_update(swr, network.river, Δt, inflow_wb, doy, update_h)
+        update(sw, swr, network, Δt)
+
+        if t + Δt > swr.Δt
+            Δt = swr.Δt - t
+        end
+        t = t + Δt
+
+        # update of h takes place in land part of shallow water, copy to river domain
+        for i = 1:swr.n
+            swr.h[i] = sw.h[inds_riv[i]]
+            swr.h_av[i] += swr.h[i] * Δt
+        end
+    end
+    swr.q_av ./= swr.Δt
+    swr.h_av ./= swr.Δt
+end
+
+function update(sw::ShallowWaterLand, swr::ShallowWaterRiver, network, Δt)
+
+    indices = network.land.staggered_indices
+    inds_riv = network.land.index_river
+
+    @unpack nodes_at_link, links_at_node = network.river
+
+    sw.qx0 .= sw.qx
+    sw.qy0 .= sw.qy
+
+    # update qx
+    @threads for i = 1:sw.n
+        yu = indices.yu[i]
+        yd = indices.yd[i]
+        xu = indices.xu[i]
+        xd = indices.xd[i]
```

Comment:
It isn't quite clear to me how we iterate over space within a timestep for overland flow, and what the dependencies are like. Why is at-threads safe here? It would also be good to know if this at-threads helps us at all for the average model, otherwise we may as well remove it.

Reply:
There are no dependencies within a timestep, so the numerical solution is very suitable for threading (we do the same for river flow).